packages/services/src/view-service/util/parallel.ts

```diff
@@ -0,0 +1,77 @@
+import { JsonObject, JsonValue } from '@bufbuild/protobuf';
+import { FullViewingKey } from '@penumbra-zone/protobuf/penumbra/core/keys/v1/keys_pb';
+import {
+  Action,
+  TransactionPlan,
+  WitnessData,
+} from '@penumbra-zone/protobuf/penumbra/core/transaction/v1/transaction_pb';
+import { WorkerBuildAction } from './build-action-worker.js';
+import { ConnectError } from '@connectrpc/connect';
+
+/**
+ * Build actions in parallel by launching a worker for each action in the plan.
+ * @returns An individually-promised list of build results.
+ */
+export const launchActionWorkers = (
```

Comment:
**comment:** for external reviewers, this delineates some of the design considerations based on offline conversations. 

The service worker is a single-threaded runtime, and objective is to lift the block processor to operate in an offscreen context, preventing it from blocking the service worker.  The possible approaches to achieve this are:

1. **target standard dom**
2. **require offscreen client to be provided**

-------------------------------------------------------------------------------------------------------------------
**target standard dom**

previously, the services package was responsible for managing the offscreen state by relying on chrome's [offscreen API](https://developer.chrome.com/docs/extensions/reference/api/offscreen). The services package was _refactored_ to abandon the offscreen API entirely, instead using standard DOM (built-in [Worker API](https://developer.mozilla.org/en-US/docs/Web/API/Worker)) to manage workers. 

This approach is more general and works outside of chrome extension limitations. If standard DOM is available, the Worker API is used directly. If not, a [polyfill](https://github.com/prax-wallet/web/pull/190) simulates the same environment (e.g., using offscreen workers). Services and block processors can directly create and manage workers using a standardized API, and there’s no dependency on chrome’s runtime or messaging. 

-------------------------------------------------------------------------------------------------------------------

**require offscreen client to be provided (via context, or constructor)**

this approach abstracts the worker-launching mechanism so that services don’t manage workers directly. Instead, workers are launched through a generalized interface, which could be either a standard DOM worker to a chrome offscreen worker. The services packages would be _agnostic_ to the specific worker implementation and lifecycle.

This would require moving the offscreen client into the handler context for services and pass as a parameter for the block processor on initialization, so they can use the same interface to launch and manage offscreen. 

In comparison to the first approach, this serves as a generalized worker launching interface provided externally to launch workers, where you're calling a  generic interface method rather than a worker constructor (standard DOM). 

-------------------------------------------------------------------------------------------------------------------

We opted for the first approach, as the difference between the two is minimal in terms of the actual implementation effort.